### PR TITLE
[Eppeta 17] Configuration, EvaluationAPI, Fetch & React Page

### PR DESCRIPTION
**Pull request by StevenM0**

### CODE REVIEW

File: src/reactapp/src/pages/EvaluationForm.js [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+  useEffect(() => {
+    fetchEvaluationObjectives();
+  }, []);
+
+  // Retrieve evaluation objectives from API
+  // Needed to fetch to localhost:7065 with the full URL to avoid CORS issues?
+  const fetchEvaluationObjectives = async () => {
+    try {
+      const response = await fetch("https://localhost:7065/api/EvaluationApi");
```
Comment (stephenfuqua) [inline src/reactapp/src/pages/EvaluationForm.js]: Please add a function to the FetchHelpers.js for a GET request, if there isn't one yet. Then you can see the other examples of how to add the authorization header to the request. Also, you'll see a note about eventually moving away from the hard-coded localhost base URL.

File: src/reactapp/package.json [hunk 2]
```diff
@@ -19,6 +19,7 @@
     "web-vitals": "^2.1.4"
   },
   "scripts": {
+        "devStart": "set HTTPS=true&&react-scripts start",
```
Comment (stephenfuqua) [inline src/reactapp/package.json]: Appears to be the same as the `start` command below.

File: src/package.json
Comment (stephenfuqua) [inline src/package.json:1]: Is this file really necessary, since the Node.js project is under `reactapp`?

File: src/package-lock.json
Comment (stephenfuqua) [inline src/package-lock.json:1]: Is this file really necessary, since the Node.js project is under `reactapp`?

File: src/reactapp/public/index.html [hunk 3]
```diff
@@ -24,7 +23,7 @@
       work correctly both with client-side routing and a non-root public URL.
       Learn how to configure a non-root public URL by running `npm run build`.
     -->
-    <title>React App</title>
+    <title>Candidate Evaluation Tracker</title>
```
Comment (stephenfuqua) [inline src/reactapp/public/index.html]: Need to rebase from `main`

File: src/reactapp/src/pages/Main.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    ];
+
+    return (
+        <Stack spacing={8}>
+            <Stack align={"center"}>
+                <Heading mt="24px" fontSize={"5xl"}>
+                    Evaluations
+                </Heading>
+            </Stack>
+            <TableContainer style={{ maxWidth: "100%" }} className="responsiveTableContainer">
```
Comment (stephenfuqua) [inline src/reactapp/src/pages/Main.js]: 100% maxwidth usually doesn't look very good - better to have a little bit of "white space" (even if it is actually another color) around the edges. Please try 95%, or add left and right margins. I would try `5rem` and see how that looks, then adjust up or down for a more please effect.

File: src/reactapp/src/pages/Main.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+      <Table variant="simple" size="lg" className="responsiveTable">
+        <Thead>
+          <Tr>
+            <Th bg="blue.500" color="white">
```
Comment (stephenfuqua) [inline src/reactapp/src/pages/Main.js]: These colors should only be applied in the CSS file on `.responsiveTable th { }`.

File: src/reactapp/src/pages/Main.js [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+            </Tr>
+          ))}
+        </Tbody>
+      </Table>
+    </TableContainer>
+            <Box mt="5" textAlign="center" mb="10">
+                <Button
+                    onClick={() => (window.location.href = "/new")}
+                    bg={"blue.400"}
+                    color={"white"}
```
Comment (stephenfuqua) [inline src/reactapp/src/pages/Main.js]: Another good opportunity to apply universal button styling in CSS, so that you don't need `bg`, `color`, or `_hover` here in the code.

File: src/reactapp/src/pages/NewEvaluation.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                <Box
+                    rounded={"lg"}
+                    bg={useColorModeValue("white", "gray.700")}
+                    boxShadow={"lg"}
+                    alignItems={"center"}
+                    w="100%"
+                    p={8}
+                >
+                    <Stack spacing={4}>
+                        <FormControl>
```
Comment (stephenfuqua) [inline src/reactapp/src/pages/NewEvaluation.js:36]: When you are ready to bind this form to real data, you'll want to convert it to a Formik form.

File: src/webapi/Controllers/EvaluationApiController.cs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+using Microsoft.AspNetCore.Mvc;
+using EdFi.OdsApi.Sdk.Client;
+using EdFi.OdsApi.SdkClient;
+using eppeta.webapi.Service;
+
+namespace webapi.Controllers
+{
+    [ApiController]
+    [Route("api/[controller]")]
+    public class EvaluationApiController : ControllerBase
```
Comment (stephenfuqua) [inline src/webapi/Controllers/EvaluationApiController.cs]: "API" is already understood, so please rename as `EvaluationController`.

File: src/webapi/Controllers/EvaluationApiController.cs [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+        [HttpGet("configuration")]
+        public Configuration GetAuthenticatedConfiguration()
+        {
+            var authenticatedConfiguration = _service.GetAuthenticatedConfiguration();
+            return authenticatedConfiguration;
+        }
+
+        // GET: api/EvaluationApi
+        // The GetEvaluation method is slow due to the need to retrieve configuration first
```
Comment (stephenfuqua) [inline src/webapi/Controllers/EvaluationApiController.cs]: In the future we can optimize this through local caching. For right now, it is fine that the method is a little slow.

File: src/webapi/Controllers/EvaluationApiController.cs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+            try
+            {
+                // Get the authenticated configuration using the injected service
+                var authenticatedConfiguration = _service.GetAuthenticatedConfiguration();
+
+                // Get Evaluation Objectives
+                var apiInstance = new EvaluationObjectivesApi(authenticatedConfiguration);
+                apiInstance.Configuration.DefaultHeaders.Add("Content-Type", "application/json");
+
+                var evaluationObjectivesWithHttpInfo = apiInstance.GetEvaluationObjectivesWithHttpInfo(limit: 25, offset: 0, totalCount: true);
```
Comment (stephenfuqua) [inline src/webapi/Controllers/EvaluationApiController.cs]: What if there are more than 25? You don't have to solve for that immediately in this ticket - but if you _don't_ solve for it right now, then please create a ticket in Jira to "Add Paging Support in the Evaluation List", and then add a comment above this line, something like `// TODO: add paging support in {JIRA TICKET ID}`.

File: ssl/dhparam.pem
Comment (stephenfuqua) [inline ssl/dhparam.pem:1]: This file should not be checked in, same for the other SSL files. In fact, they shouldn't even exist in this directory. Please delete the root-level `ssl` directory.

File: src/webapi/Service/AuthenticationConfigurationService.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+            System.Net.ServicePointManager.SecurityProtocol |= System.Net.SecurityProtocolType.Tls11 | System.Net.SecurityProtocolType.Tls12;
+
+            // TokenRetriever makes the oauth calls. It has RestSharp dependency, install via NuGet
+            var tokenRetriever = new TokenRetriever(oauthUrl, clientKey, clientSecret);
+
+            // Plug Oauth access token. Tokens will need to be refreshed when they expire
+            var configuration = new Configuration()
+            {
+                AccessToken = tokenRetriever.ObtainNewBearerToken(),
+                BasePath = "https://localhost:443/api/data/v3/"
```
Comment (stephenfuqua) [inline src/webapi/Service/AuthenticationConfigurationService.cs]: For this URL, please create a new value in the `appsettings.json` file, and in the `AppSettings.cs` singleton. This will help us avoid hard-coding.

File: src/webapi/Service/AuthenticationConfigurationService.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+            this.oauthUrl = oauthUrl;
+            this.clientKey = clientKey;
+            this.clientSecret = clientSecret;
+        }
+
+        public Configuration GetAuthenticatedConfiguration()
+        {
+            // Trust all SSL certs -- needed unless signed SSL certificates are configured.
+            System.Net.ServicePointManager.ServerCertificateValidationCallback =
+                ((sender, certificate, chain, sslPolicyErrors) => true);
```
Comment (stephenfuqua) [inline src/webapi/Service/AuthenticationConfigurationService.cs]: This is option needs to be controlled by configuration - not good to always accept the SSL certificates. Need to add something to the appsettings file and the `AppSettings` singleton.

File: src/webapi/Service/AuthenticationConfigurationService.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+            // Explicitly configures outgoing network calls to use the latest version of TLS where possible.
+            // Due to our reliance on some older libraries, the.NET framework won't necessarily default
+            // to the latest unless we explicitly request it. Some hosting environments will not allow older versions
+            // of TLS, and thus calls can fail without this extra configuration.
+            System.Net.ServicePointManager.SecurityProtocol |= System.Net.SecurityProtocolType.Tls11 | System.Net.SecurityProtocolType.Tls12;
```
Comment (stephenfuqua) [inline src/webapi/Service/AuthenticationConfigurationService.cs]: This, and the SSL certificate trusting above, only needs to run once at application startup. It would be fine to move these to `Program.cs`.

File: src/webapi/Service/AuthenticationConfigurationService.cs [hunk 14]
```diff
@@ -0,0 +1,44 @@
+using EdFi.OdsApi.Sdk.Client;
+using EdFi.OdsApi.SdkClient;
+
+namespace eppeta.webapi.Service
+{
+    public class AuthenticationConfigurationService : IAuthenticationConfigurationService
```
Comment (stephenfuqua) [inline src/webapi/Service/AuthenticationConfigurationService.cs]: This class seems pretty good and you've done a nice job with dependency injection.

There is potential confusion here for future developers. There are two uses of authentication in this application: end-users logging in, and the service connecting to the Ed-Fi ODS/API. It may be useful to rename this class to help make it more clear that this serves the second purpose.

**Review by stephenfuqua — COMMENTED** [review]
Making good progress here and I like what I'm seeing. I've offered a lot of individual comments to help take next steps in improving the code design, readability, and simplicity.

File: .vscode/launch.json [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            // Use IntelliSense to find out which attributes exist for C# debugging
+            // Use hover for the description of the existing attributes
+            // For further information visit https://github.com/dotnet/vscode-csharp/blob/main/debugger-launchjson.md
+            "name": ".NET Core Launch (web)",
+            "type": "coreclr",
+            "request": "launch",
+            "preLaunchTask": "build",
+            // If you have changed target frameworks, make sure to update the program path.
+            "program": "${workspaceFolder}/src/webapi/bin/Debug/net7.0/webapi.dll",
```
Comment (stephenfuqua) [inline .vscode/launch.json]: This project is using .NET 6, not .NET 7. Originally it was using 7, but we have a policy to avoid it because it is not the "long term support" version of .NET. In other words, we'll be forced to update it next year. But if we use .NET 6, we won't be forced to update until 2025.

Comment (stephenfuqua) [reply .vscode/launch.json]: Also change `webapi.dll` --> `eppeta.webapi.dll`

File: src/webapi/Controllers/EvaluationApiController.cs [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+        public Configuration GetAuthenticatedConfiguration()
+        {
+            var authenticatedConfiguration = _service.GetAuthenticatedConfiguration();
+            return authenticatedConfiguration;
+        }
+
+        // GET: api/EvaluationApi
+        // The GetEvaluation method is slow due to the need to retrieve configuration first
+        [HttpGet]
+        public async Task<ActionResult<IEnumerable<TpdmEvaluationObjective>>> GetEvaluation()
```
Comment (stephenfuqua) [inline src/webapi/Controllers/EvaluationApiController.cs]: Should this function be called `GetEvaluationObjectives`?

File: src/webapi/Controllers/EvaluationApiController.cs (hunk 10 above)
Comment (stephenfuqua) [reply src/webapi/Controllers/EvaluationApiController.cs]: After discussing further, it looks like paging may not be necessary because there aren't enough Evaluation Objectives to require it. Therefore ignore the comment above.

@stevenarnold-ed-fi please confirm; if you think that there will be more than 25 evaluation objectives at some point, then can you create another ticket to support getting more objectives?

File: src/EdFi_Teacher_Evaluation.sln [hunk 17]
```diff
@@ -23,6 +23,8 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "SQL", "SQL", "{47708811-A8F
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "eppeta.webapi.unitTests", "eppeta.webapi.unitTests\eppeta.webapi.unitTests.csproj", "{38A32D34-151D-4D27-941B-E8E3E72F19E4}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EdFi.OdsApi.Sdk", "..\..\..\..\..\..\csharp\src\EdFi.OdsApi.Sdk\EdFi.OdsApi.Sdk.csproj", "{E0EA1A72-7E1E-4C3E-95FC-6F042B9E91CA}"
```
Comment (stephenfuqua) [inline src/EdFi_Teacher_Evaluation.sln]: When I ran `dotnet build` just now I got an error:
```
C:\Program Files\dotnet\sdk\7.0.302\NuGet.targets(398,5): error MSB3202: The project file "C:\csharp\src\EdFi.OdsApi.Sd
k\EdFi.OdsApi.Sdk.csproj" was not found. [C:\source\exchange\EvaluationTracker\src\EdFi_Teacher_Evaluation.sln]
```

And now it is clear why: you accidentally forgot to copy the SDK into the repository, as we see with the long string of `..\` characters above 😆 .

File: src/webapi/AppSettings.cs [hunk 18]
```diff
@@ -54,6 +55,30 @@ public static string[] AllowedOrigins
...
+            // Trust all SSL certs -- needed unless signed SSL certificates are configured.
+            System.Net.ServicePointManager.ServerCertificateValidationCallback =
+                ((sender, certificate, chain, sslPolicyErrors) => true);
+
+            // Due to our reliance on some older libraries, the.NET framework won't necessarily default
+            // to the latest unless we explicitly request it. Some hosting environments will not allow older versions
+            // of TLS, and thus calls can fail without this extra configuration.
+            System.Net.ServicePointManager.SecurityProtocol |= System.Net.SecurityProtocolType.Tls11 | System.Net.SecurityProtocolType.Tls12;
+        }
+    }
```
Comment (stephenfuqua) [inline src/webapi/AppSettings.cs:81]: For right now, let's leave this code alone. My comments are just educational.

There is a concept called the Single Responsibility Principle (SRP), sometimes also called "do just one thing". I have a [blog post](https://blog.safnet.com/best-practices-tdd-oo/solid-testing) that addresses SRP and related topics, which you might find interesting.

In this case, the `AppSettings` file's _single responsibility_ should be to retrieve configuration values. This function does more than that: it also makes modifications to the security.

In the long term, it would be better for those modifications to occur in `Program.cs`, whose single responsibility is to setup the application infrastructure. Then in `AppSettings` you would just have:

```
    public static bool TrustAllSSLCerts
    {
        get
        {
            return GetInstance()._configuration.GetValue<bool>("TrustAllSSLCerts");
        }
    }
```

And then in `Program.cs` you would have

```
   if (AppSettings.TrustAllSSLCerts)
        {
            // Trust all SSL certs -- needed unless signed SSL certificates are configured.
            System.Net.ServicePointManager.ServerCertificateValidationCallback =
                ((sender, certificate, chain, sslPolicyErrors) => true);

            // Due to our reliance on some older libraries, the.NET framework won't necessarily default
            // to the latest unless we explicitly request it. Some hosting environments will not allow older versions
            // of TLS, and thus calls can fail without this extra configuration.
            System.Net.ServicePointManager.SecurityProtocol |= System.Net.SecurityProtocolType.Tls11 | System.Net.SecurityProtocolType.Tls12;
        }
```

You can refactor this in your next ticket. Let's leave it alone right now.

File: src/webapi/appsettings.json [hunk 19]
```diff
@@ -21,5 +21,7 @@
     "NewUsersAreAdministrators": false,
     "RequireHttps": true
   },
-  "CorsAllowedOrigins": ""
+  "CorsAllowedOrigins": "",
+  "BasePath": "https://localhost:443/api/data/v3/",
+  "TrustAllSSLCerts": true
```
Comment (stephenfuqua) [inline src/webapi/appsettings.json]: The `appsettings.json` file should usually contain universal defaults. Both of the lines above are inappropriate for default values. We can override with _development machine defaults_ by adding them to `appsettings.Development.json`. For this file, we should simply have:

```json
  "BasePath": "",
  "TrustAllSSLCerts": false
```

**Review by stephenfuqua — CHANGES_REQUESTED** [review]
To move forward, please add the SDK project to the repository directory, and adjust the appsettings file as described in another comment.

File: EdFi.OdsApi.Sdk/Apis.All/CandidatesApi.cs [hunk 20]
```diff
@@ -0,0 +1,2303 @@
+/*
```
Comment (stephenfuqua) [inline EdFi.OdsApi.Sdk/Apis.All/CandidatesApi.cs:1]:
```suggestion
// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.
```

File: EdFi.OdsApi.Sdk/Apis.All/EvaluationObjectivesApi.cs [hunk 21]
```diff
@@ -0,0 +1,2111 @@
+/*
```
Comment (stephenfuqua) [inline EdFi.OdsApi.Sdk/Apis.All/EvaluationObjectivesApi.cs:1]:
```suggestion
// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.
```

File: src/webapi/Program.cs [hunk 22]
```diff
@@ -27,17 +28,24 @@ private static void Main(string[] args)
...
-
         ConfigureDatabaseConnection(builder);
         ConfigureLocalIdentityProvider(builder.Services);
         ConfigureQuartz(builder.Services);
         ConfigureSwaggerUIServices(builder.Services);
         ConfigureAspNetAuth(builder.Services);
 
+        // Add authentication configuration service to the container.
+        builder.Services.AddScoped<IODSAPIAuthenticationConfigurationService>(
+            provider => new ODSAPIAuthenticationConfigurationService("https://localhost:443/api/", "populated", "populatedSecret")
```
Comment (stevenarnold-ed-fi) [inline src/webapi/Program.cs:45]: Per other comments, this should also come from configuration

**Review by stephenfuqua — APPROVED** [review]

**Review by stevenarnold-ed-fi — APPROVED** [review]
